std/datetime/stopwatch.d

```diff
@@ -0,0 +1,425 @@
+
+        Duration t5 = sw.peek();
+        assert(t5 > t4);
+
+        // If stopping or resetting the StopWatch is not required, then
+        // MonoTime can easily be used by itself without StopWatch.
+        auto before = MonoTime.currTime;
+        // do stuff...
+        auto timeElapsed = MonoTime.currTime - before;
+    }
+
+    /++
+        Constructs a StopWatch. Whether it starts immediately depends on the
+        $(LREF AutoStart) argument.
+
+        If $(D StopWatch.init) is used, then the constructed StopWatch isn't
+        running (and can't be, since no constructor ran).
+      +/
+    this(AutoStart autostart) @safe nothrow @nogc
+    {
+        if (autostart)
+            start();
+    }
+
+    ///
+    @system nothrow @nogc unittest
+    {
+        import core.thread : Thread;
+
+        {
+            auto sw = StopWatch(AutoStart.yes);
+            assert(sw.running);
+            Thread.sleep(usecs(1));
+            assert(sw.peek() > Duration.zero);
+        }
+        {
+            auto sw = StopWatch(AutoStart.no);
+            assert(!sw.running);
+            Thread.sleep(usecs(1));
+            assert(sw.peek() == Duration.zero);
+        }
+        {
+            StopWatch sw;
+            assert(!sw.running);
+            Thread.sleep(usecs(1));
+            assert(sw.peek() == Duration.zero);
+        }
+
+        assert(StopWatch.init == StopWatch(AutoStart.no));
+        assert(StopWatch.init != StopWatch(AutoStart.yes));
+    }
+
+
+    /++
+       Resets the StopWatch.
+
+       The StopWatch can be reset while it's running, and resetting it while
+       it's running will not cause it to stop.
+      +/
+    void reset() @safe nothrow @nogc
+    {
+        if (_running)
+            _timeStarted = MonoTime.currTime;
+        _ticksElapsed = 0;
+    }
+
+    ///
+    @system nothrow @nogc unittest
+    {
+        import core.thread : Thread;
+
+        auto sw = StopWatch(AutoStart.yes);
+        Thread.sleep(usecs(1));
+        sw.stop();
+        assert(sw.peek() > Duration.zero);
+        sw.reset();
+        assert(sw.peek() == Duration.zero);
+    }
+
+    @system nothrow @nogc unittest
+    {
+        import core.thread : Thread;
+
+        auto sw = StopWatch(AutoStart.yes);
+        Thread.sleep(msecs(1));
+        assert(sw.peek() > msecs(1));
+        immutable before = MonoTime.currTime;
+
+        // Just in case the system clock is slow enough or the system is fast
+        // enough for the call to MonoTime.currTime inside of reset to get
+        // the same that we just got by calling MonoTime.currTime.
+        Thread.sleep(usecs(1));
+
+        sw.reset();
+        assert(sw.peek() < msecs(1));
+        assert(sw._timeStarted > before);
+        assert(sw._timeStarted < MonoTime.currTime);
+    }
+
+
+    /++
+       Starts the StopWatch.
+
+       start should not be called if the StopWatch is already running.
+      +/
+    void start() @safe nothrow @nogc
+    in { assert(!running); }
```

Comment:
There was a longer discussion this summer on which a majority agreed that an `assert` statement should have a message even if it's trivial as it greatly helps the user in figuring out what went wrong instead of the bare segfault.